### PR TITLE
i#6524: Fix udivmoddi4.c 32-bit linux build fail and semantics

### PR DESCRIPTION
**Pull request by xdje42**
For the targets we care about LIBGCC2_UNITS_PER_WORD == 4 (SImode in GCC parlance).
Thus a double word is 8 bytes (DImode in GCC parlance).

Previously, __moddi3 was using a mix of word and double word widths,
causing a compilation failure on 32 bit linux with the Debian 13.2.0-4 compiler.
The semantics were also incorrect: inputs and outputs should be 64 bits for the
hosts we care about (x86, aarch, 32 and 64 bit, and riscv64).

Fixes: #6524

### CODE REVIEW

File: third_party/libgcc/udivmoddi4.c [hunk 1]
```diff
@@ -23,11 +23,15 @@ a copy of the GCC Runtime Library Exception along with this program;
 see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
```
Comment (abhinav92003) [inline third_party/libgcc/udivmoddi4.c:23]: Following the example set by PR #2605, I believe we also need to update third_party/libgcc/README.dynamorio.

File: third_party/libgcc/udivmoddi4.c [hunk 2]
```diff
@@ -23,11 +23,15 @@ a copy of the GCC Runtime Library Exception along with this program;
...
 
-/* This is extracted from gcc's libgcc/libgcc2.c with these typedefs added: */
-typedef short Wtype;
-typedef int DWtype;
-typedef unsigned int UWtype;
-typedef unsigned long long UDWtype;
+#include <stdint.h>
+
+/* This is extracted from gcc's libgcc/libgcc2.c with these typedefs added.
+   Note that for the targets we care about LIBGCC2_UNITS_PER_WORD == 4.
```
Comment (abhinav92003) [inline third_party/libgcc/udivmoddi4.c:29]: Would be nicer for clarity to mention the targets, so that we know if/when this needs to be updated.

**Review by abhinav92003 — APPROVED** [review]

File: third_party/libgcc/udivmoddi4.c (hunk 1 above)
Comment (xdje42) [reply third_party/libgcc/udivmoddi4.c:23]: Done

File: third_party/libgcc/udivmoddi4.c (hunk 2 above)
Comment (xdje42) [reply third_party/libgcc/udivmoddi4.c]: Done

**Pull request merged**